pkg/reconciler/managed/metrics.go

```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2024 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package managed
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	corev1 "k8s.io/api/core/v1"
+	kmetrics "k8s.io/component-base/metrics"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+)
+
+const subSystem = "crossplane"
+
+// MetricRecorder records the managed resource metrics.
+type MetricRecorder interface { //nolint:interfacebloat // The first two methods are coming from Prometheus
+	Describe(ch chan<- *prometheus.Desc)
+	Collect(ch chan<- prometheus.Metric)
+
+	recordUnchanged(name string)
+	recordFirstTimeReconciled(managed resource.Managed)
+	recordFirstTimeReady(managed resource.Managed)
+	recordDrift(managed resource.Managed)
+	recordDeleted(managed resource.Managed)
+}
+
+// MRMetricRecorder records the lifecycle metrics of managed resources.
+type MRMetricRecorder struct {
+	firstObservation sync.Map
+	lastObservation  sync.Map
+
+	mrDetected       *prometheus.HistogramVec
+	mrFirstTimeReady *prometheus.HistogramVec
+	mrDeletion       *prometheus.HistogramVec
+	mrDrift          *prometheus.HistogramVec
+}
+
+// NewMRMetricRecorder returns a new MRMetricRecorder which records metrics for managed resources.
+func NewMRMetricRecorder() *MRMetricRecorder {
+	return &MRMetricRecorder{
+		mrDetected: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Subsystem: subSystem,
+			Name:      "managed_resource_time_to_first_reconcile_seconds",
```

Comment:
Nit: Does it make sense to make the following names consistent with each other: `managed_resource_time_to_first_reconcile_seconds`, `managed_resource_first_time_to_readiness_seconds`? `time_to_first` sounds easier to the ear, so we can make the latter `managed_resource_time_to_first_readiness_seconds`, to agree with the former. On the other hand, who are familiar with time to readiness (TTR) might find `first_time_to_readiness` more natural. In that case we can leave them as they are. Changing the former to `managed_resource_first_time_to_reconcile_seconds` sounds weird to me.